crypto/rsa/rsa_gen.c

```diff
@@ -85,62 +133,164 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
+
+        /* prime_info from 2 to |primes| -1 */
+        for (i = 2; i < primes; i++) {
+            pinfo = rsa_multip_info_new();
+            if (pinfo == NULL)
+                goto err;
+            (void)sk_RSA_PRIME_INFO_push(prime_infos, pinfo);
+        }
     }
-    if (!BN_GENCB_call(cb, 3, 0))
+
+    if (BN_copy(rsa->e, e_value) == NULL)
         goto err;
-    for (;;) {
-        do {
-            if (!BN_generate_prime_ex(rsa->q, bitsq, 0, NULL, NULL, cb))
+
+    /* generate p, q and other primes (if any) */
+    for (i = 0; i < primes; i++) {
+        if (i == 0) {
+            prime = rsa->p;
+        } else if (i == 1) {
+            prime = rsa->q;
+        } else {
+            pinfo = sk_RSA_PRIME_INFO_value(prime_infos, i - 2);
+            prime = pinfo->r;
+        }
+
+        for (;;) {
+ redo:
+            if (!BN_generate_prime_ex(prime, bitsr[i] + adj, 0, NULL, NULL, cb))
+                goto err;
+            /*
+             * prime should not be equal to p, q, r_3...
+             * (those primes prior to this one)
+             */
+            {
+                int j = 0;
+                BIGNUM *prev_prime = NULL;
+
+                for (j = 0; j < i; j++) {
+                    if (j == 0)
+                        prev_prime = rsa->p;
+                    else if (j == 1)
+                        prev_prime = rsa->q;
+                    else
+                        prev_prime = sk_RSA_PRIME_INFO_value(prime_infos,
+                                                             j - 2)->r;
+
+                    if (!BN_cmp(prime, prev_prime)) {
+                        goto redo;
+                    }
+                }
+            }
+            if (!BN_sub(r2, prime, BN_value_one()))
+                goto err;
+            if (!BN_gcd(r1, r2, rsa->e, ctx))
                 goto err;
-        } while (BN_cmp(rsa->p, rsa->q) == 0);
-        if (!BN_sub(r2, rsa->q, BN_value_one()))
+            if (BN_is_one(r1))
+                break;
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
+        }
+
+        bitse += bitsr[i];
+
+        /* calculate n immediately to see if it's sufficient */
+        if (i == 1) {
+            /* we get at least 2 primes */
+            if (!BN_mul(r1, rsa->p, rsa->q, ctx))
+                goto err;
+        } else if (i != 0) {
+            /* modulus n = p * q * r_3 * r_4 ... */
+            if (!BN_mul(r1, rsa->n, prime, ctx))
+                goto err;
+        } else {
+            /* i == 0, do nothing */
+            if (!BN_GENCB_call(cb, 3, i))
+                goto err;
+            continue;
+        }
+        nl = BN_num_bits(r1);
+        /*
+         * if |r1|, product of factors so far, is not long enough,
+         * re-generate the last prime
+         *
+         * NOTE: This actually can't happen in two-prime case, because of
+         * the way factors are generated.
+         */
+        if (nl != bitse) {
```

Comment:
I think two-prime RSA modules will have the requested number of bits,
but the first 3 bits will never be 1 0 0, because the primes numbers
begin with 1 1 thus they are in the range 3/4 * 2^(n/2) .. 2 * 2^(n/2)
and therefore the product is in the range 9/16 * 2^n .. 2 n.
But with this test 1/8 of all generated RSA modules will begin
with 1 0 0 can be spotted by looking at the first 3 bits of the public exponent.

Reply:
So I would suggest to check the first 4 mantissa bits need to be 0x9..0xF
if they are outside, repeat search.

Reply:
> with 1 0 0 0 can be spotted by looking at the first 4 bits of the public exponent.

By public exponent, you probably meant 'the product of primes so far' ?

Reply:
And I am sorry I don't understand why with this test, 1/8 of all generated RSA modulus begin with `1000`, would you please explain a bit more? Thanks! (I think the discussion is based on two-prime RSA case, isn't it?)

Reply:
> By public exponent, you probably meant 'the product of primes so far' (or the 'modulus')?

Yes, in the case of 2-prime RSA the module N will always start with 0x9..0xF, never with 0x8.
This number is embedded in the public certificate, and can be extracted trivially.

But in the case of n-prime RSA there is a likelihood that the module can have the
requested number of bits, but start with 0x8.

Certificates using modules that start with 0x8 can be sure to be multi-prime RSA.
Once you know they are constructed of small prime numbers other factorization
methods than the General Prime Number Sieve could be used, most importantly
the Elliptic Curve Factorization Method.

You would not want to let anybody know for sure that a given RSA public key is
in fact a multi-prime RSA.

Reply:
I expect 1/8 of all generated multi-prime RSA modules to begin with 0x8
because the generated k-prime RSA numbers will range from (3/4)^k * 2^N .. 2^N
in the first iteration, but the check here re-generates every module less than 0.5 *2^N
so I expect the modules to be evenly distributed from 0.5*2^N..2^N.

Thus 1 out of 8 modules start with 0x8.

Reply:
> Certificates using modules that start with 0x8 can be sure to be multi-prime RSA.

Ahh, this is a very good point. Thanks for pointing out and I will make a test for the first 4 bits here.

Reply:
> You would not want to let anybody know for sure that a given RSA public key is
in fact a multi-prime RSA.

It's not something you should assume can be concealed, not fact itself, nor amount of factors. It's inevitable that they would figure it out. By simply measuring the time it took you to perform private key operation...

Reply:
Yes, they would still need to set that time into relation to the time it takes you to do full
2-prime RSA private op.
But i think there *are* people doing offline scans for weak RSA keys.

Reply:
Yes. But it's still valuable to remove such a 'straightforward' clue in the certificate, to stop *some* people but not *all* the people from knowing this is an multi-prime key. This is especially useful in some *real* environments...